interactive/dashboard/flow_diagram.py

```diff
@@ -119,153 +102,794 @@ def name_included_reactions(included_reactions, reactions_json):
+                elif (isSpeciesInReaction(reaction_data, product_name)
+                      or isSpeciesInReaction(reactant_data, product_name)):
+                    product_yield = product_yield['yield']
+                    name = str(speciesFromReaction)+"__TO__"+str(product_name)
+                    quantities.update(
+                        {name: product_yield})
+                    if str(speciesFromReaction) in reactions_nodes:
+                        quan = total_quantites.get(product_name, 0)
+                        new_val = quan + product_yield
+                        total_quantites.update(
+                            {str(product_name): new_val})
+                        tmp = str(product_yield) + str(product_name)+"_"
+                        products_string += tmp
+                        products_string = products_string.replace(".0"+str(
+                            product_name), str(product_name))
+            products_string = products_string[:-1]
+            print("* pushing to hover:", reactants_string, "->", products_string)
+            reaction_names_on_hover.update(
+                {speciesFromReaction: reactants_string+"->"+products_string})
+        i += 1
+    return quantities, total_quantites, reaction_names_on_hover
+
+
+def CalculateEdgesAndNodes(reactions, species, scaledLineWeights,
+                           blockedSpecies):
+    print("[7/7] Creating edges and nodes...")
+    species = {}
+    edges = {}
+    reactions = {}
+
+    for weight in scaledLineWeights:
+
+        fromElement = weight[0].split('__TO__')[0].replace("__TO__", "")
+        ToElement = weight[0].split('__TO__')[1].replace("__TO__", "")
+        lineWidth = float(weight[1])
+
+        if "->" in fromElement:
+            # reaction -> product
+            edge = (beautifyReaction(fromElement), ToElement, lineWidth)
+            edges.update({edge: {}})
+            if species not in blockedSpecies:
+                species.update({ToElement: {}})
+            reactions.update({beautifyReaction(fromElement): {}})
+        else:
+            # reactant -> reaction
+            edge = (fromElement, beautifyReaction(ToElement), lineWidth)
+            edges.update({edge: {}})
+            if species not in blockedSpecies:
+                species.update({fromElement: {}})
+            reactions.update({beautifyReaction(ToElement): {}})
+    return {'edges': list(edges), 'species_nodes': list(species),
+            'reaction_nodes': list(reactions)}
+
 
 def createLegend():
     x = -300
     y = -250
     legend_nodes = [
-       'Element', 'Reaction'
+        'Element', 'Reaction'
     ]
     return legend_nodes
+
+
 # parent function for generating flow diagram
+def getReactName(reaction_names_on_hover, x):
+    name = ""
+    try:
+        name = beautifyReaction(reaction_names_on_hover[unbeautifyReaction(x)])
+    except KeyError:
+        name = x
+    return name
 def generate_flow_diagram(request_dict):
```

Comment:
please provide a description for what this function is doing. Also, it is 250 lines long - please consider refactoring this into smaller, focused functions